src/lib/markbind/src/preprocessors/componentPreprocessor.js

```diff
@@ -0,0 +1,394 @@
+  if (fileExistsNode) return fileExistsNode;
+
+  // optional includes of whole files have been handled,
+  // but segments still need to be processed further down
+  if (isOptional && !hash) {
+    delete element.attribs.optional;
+  }
+
+  // TODO do we need boilerplateIncludeSrc?
+  if (boilerplateFilePath) {
+    parser.boilerplateIncludeSrc.push({ from: context.cwf, to: boilerplateFilePath });
+  }
+
+  const isInline = _.has(element.attribs, 'inline');
+  const isTrim = _.has(element.attribs, 'trim');
+  const isDynamic = _.has(element.attribs, 'dynamic');
+
+  element.name = isInline ? 'span' : 'div';
+  element.attribs[ATTRIB_INCLUDE_PATH] = filePath;
+
+  // Use a 'plain' panel for a dynamic include
+  if (isDynamic) return _preprocessDynamicInclude(element, context, parser, hash, filePath, actualFilePath);
+
+  // No need to process url contents
+  if (isUrl) return element;
+
+  parser.staticIncludeSrc.push({ from: context.cwf, to: actualFilePath });
+
+  const isIncludeSrcMd = _isHtmlIncludingMarkdown(element, context, filePath);
+
+  const { content, childContext, userDefinedVariables }
+    = parser._renderIncludeFile(actualFilePath, element, context, config, filePath);
+  childContext.source = isIncludeSrcMd ? 'md' : 'html';
+  childContext.callStack.push(context.cwf);
+  parser.extractInnerVariablesIfNotProcessed(content, childContext, config, filePath);
+
+  const innerVariables = parser.getImportedVariableMap(filePath);
+  const fileContent = nunjucks.renderString(content, { ...userDefinedVariables, ...innerVariables });
+
+  _deleteIncludeAttributes(element);
+
+  // Process sources with or without hash, retrieving and appending
+  // the appropriate children to a wrapped include element
+
+  let actualContent;
+
+  if (hash) {
+    // Keep scripts in the fileContent
+    const src = cheerio.parseHTML(fileContent, true);
+    const $ = cheerio.load(src);
+    const hashContent = $(hash).html();
+
+    actualContent = (hashContent && isTrim) ? hashContent.trim() : hashContent;
+
+    if (actualContent === null && isOptional) {
+      // Use empty content if it is optional
+      actualContent = '';
+    } else if (actualContent === null) {
+      const hashSrcWithoutHash = hash.substring(1);
+      const error = new Error(`No such segment '${hashSrcWithoutHash}' in file: ${actualFilePath}\n`
+          + `Missing reference in ${element.attribs[ATTRIB_CWF]}`);
+      parser._onError(error);
+
+      return utils.createErrorNode(element, error);
+    }
+
+    // optional includes of segments have now been handled, so delete the attribute
+    if (isOptional) delete element.attribs.optional;
+  } else {
+    actualContent = (fileContent && isTrim) ? fileContent.trim() : fileContent;
+  }
+
+  if (isIncludeSrcMd) {
+    if (context.mode === 'include') {
+      actualContent = isInline ? actualContent : `\n\n${actualContent}\n`;
+    } else {
+      // TODO do we need this branch?
+      actualContent = md.render(actualContent);
```

Comment:
A quick usage find for `_preprocess` reveals that it is never called with anything other than `context.mode = 'include'`